src/tools/rbd_mirror/group_replayer/Replayer.cc

```diff
@@ -403,11 +398,30 @@ void Replayer<I>::handle_load_local_group_snapshots(int r) {
       continue;
     }
     if (ns->state != cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY) {
+      if (ns->is_orphan()) {
+        dout(5) << "local group being force promoted" << dendl;
+        locker.unlock();
+        handle_replay_complete(0, "orphan (force promoting)");
+        return;
+      } else if (m_retry_validate_snap) {
+        m_retry_validate_snap = false;
+        locker.unlock();
+        schedule_load_group_snapshots();
+        return; // previous snap is syncing, reload local snaps to confirm its completeness.
+      }
       break;
+    } else  { // primary
+      if (it->state == cls::rbd::GROUP_SNAPSHOT_STATE_INCOMPLETE) {
+        dout(10) << "found incomplete primary group snapshot: "
+                 << it->id << dendl;
+        locker.unlock();
+        schedule_load_group_snapshots();
```

Comment:
Will this load the local snaps again or will it spend time processing the existing local group snaps list?

Reply:
I have added some logic in load_local_group_snapshots() to skip the validation if the snapshot is PRIMARY. That should get the state machine back here. Can you please check this once again.